src/udev/net/link-config.c

```diff
@@ -923,15 +931,108 @@ static int link_apply_sr_iov_config(Link *link, sd_netlink **rtnl) {
+                                                                        \
+                sd_device *device = (sd_device*) ASSERT_PTR(userdata);  \
+                const char *val;                                        \
+                char *dup;                                              \
+                int r;                                                  \
+                                                                        \
+                r = getter(device, &val);                               \
+                if (r < 0)                                              \
+                        return r;                                       \
+                                                                        \
+                dup = strdup(val);                                      \
+                if (!dup)                                               \
+                        return -ENOMEM;                                 \
+                                                                        \
+                *ret = dup;                                             \
+                return 0;                                               \
+        }
+
+DEFINE_DEVICE_SPECIFIER(net_driver, device_get_net_driver);
+DEFINE_DEVICE_SPECIFIER(sysname, sd_device_get_sysname);
+DEFINE_DEVICE_SPECIFIER(devpath, sd_device_get_devpath);
+
+static int device_printf(sd_device *device, const char *format, char **ret) {
+        assert(device);
+        assert(format);
+        assert(ret);
+
+        const Specifier table[] = {
+                /* These are subset of specifiers supported in udev-format.c. */
+                { 'd', specifier_net_driver, NULL },
+                { 'k', specifier_sysname,    NULL },
+                { 'p', specifier_devpath,    NULL },
+
+                COMMON_SYSTEM_SPECIFIERS,
+                COMMON_TMP_SPECIFIERS,
+                {}
+        };
+
+        return specifier_printf(format, SIZE_MAX, table, NULL, device, ret);
+}
+
 static int link_apply_udev_properties(Link *link, bool test) {
         LinkConfig *config;
         sd_device *device;
+        int r;
 
         assert(link);
 
         config = ASSERT_PTR(link->config);
         device = ASSERT_PTR(link->device);
 
+        /* 1. apply ImportProperty=. */
+        STRV_FOREACH(p, config->import_properties)
+                (void) udev_builtin_import_property(device, link->device_db_clone, test, *p);
+
+        /* 2. apply Property=. */
+        STRV_FOREACH(p, config->properties) {
+                _cleanup_free_ char *key = NULL, *val = NULL;
+                const char *eq;
+
+                eq = strchr(*p, '=');
+                if (!eq)
+                        continue;
+
+                key = strndup(*p, eq - *p);
+                if (!key)
+                        return log_oom();
+
+                r = device_printf(device, eq + 1, &val);
```

Comment:
so, hmm, i think so far we made the rule to apply specifiers at the time of config parsing. I think this would be the first place where we'd apply it at moment of application.

I am not sure how i feel about this shift, it would be quite an outlier.

Also, if you want device-specific specifiers to resolve maybe the right approach would be to use a udev rule rather than a .link file for this? because udev rules are a different language ans they always supported forms of expansion applied at application time...